### ISSUE TITLE
Reading Air-O2 Flow from SFM3003-300-CL

### ISSUE THREAD

**Issue by KHRAMS** [Sensirion/arduino-i2c-sfm-sf06#7]
Hi,

I have been able to read O2 flow successfully using this library using the aforementioned SFM3003-300-CL library on an Arduino Nano 33 BLE. However, when I tried to use the startAirO2ContinuousMeasurement() function to read the both a mixture-flow from human breath (as well as the O2 concentration within this mixture), it started to give me large negative values for flow. Further, the last nine bits of the status word did not change to match a changing level of O2 concentration as they should according to the datasheet. This may be entirely due to my lack of understanding, so it would be much appreciated to get some clarity on this.

The following code is what I am currently using (note for others who use a Nano that I had to change the name of the variable "error" as it is a keyword in the Mbed OS driver commonly used on Arduino Nano systems):

```
// #include <Arduino.h>
#include <SensirionI2cSfmSf06.h>
#include <stdlib.h>
#include <stdint.h>

#include <Wire.h>

SensirionI2cSfmSf06 sensor;

static char errMessage[64];
static int16_t err;


void print_byte_array(uint8_t* array, uint16_t len) {
  uint16_t i = 0;
  Serial.print("0x");
  for (; i < len; i++) {
    Serial.print(array[i], HEX);
  }
}




void setup() {

  Serial.begin(115200);
  while (!Serial) {
    delay(100);
  }
  Wire.begin();
  sensor.begin(Wire, SFM3003_I2C_ADDR_28);

  sensor.stopContinuousMeasurement();
  delay(100);
  uint32_t productIdentifier = 0;
  uint8_t serialNumber[8] = { 0 };
  err = sensor.readProductIdentifier(productIdentifier, serialNumber, 8);
  if (err != NO_ERROR) {
    Serial.print("Error trying to execute readProductIdentifier(): ");
    errorToString(err, errMessage, sizeof errMessage);
    Serial.println(errMessage);
    return;
  }
  Serial.print("productIdentifier: ");
  Serial.print(productIdentifier);
  Serial.print("\t");
  Serial.print("serialNumber: ");
  print_byte_array(serialNumber, 8);
  Serial.println();

  
  uint16_t o2_vol_frac = 21; // in %, assuming 21% volume O2 in ambient air mixture.
  err = sensor.startAirO2ContinuousMeasurement(o2_vol_frac);
  if (err != NO_ERROR) {
    Serial.print(
      "Error trying to execute startO2ContinuousMeasurement(): ");
    errorToString(err, errMessage, sizeof errMessage);
    Serial.println(errMessage);
    return;
  }
}

void loop() {
 
  float aFlow = 0.0;
  float aTemperature = 0.0;
  uint16_t aStatusWord = 0u;
  delay(100);
  err = sensor.readMeasurementData(aFlow, aTemperature, aStatusWord);
  if (err != NO_ERROR) {
    Serial.print("Error trying to execute readMeasurementData(): ");
    errorToString(err, errMessage, sizeof errMessage);
    Serial.println(errMessage);
    return;
  }
  Serial.print("aFlow: ");
  Serial.println(aFlow);
  Serial.print("aTemperature: ");
  Serial.println(aTemperature);
  Serial.print("aStatusWord: ");
  Serial.println(aStatusWord);
}
```

The resulting output looks something like :

```
aFlow: -24583.00
aTemperature: 27.95
aStatusWord: 26645
```
which seems entirely incorrect. I am assuming that readFlowMeasurement is accurately using the scale factors and such due to its prior performance, which leaves either my use of the O2 volume input parameter or my use of the read function itself at fault.

Thank you for your help!

**Comment by quentin-cy** [comment]
Hello @KHRAMS ,

Interesting issue you seem to have here. I will try to reproduce your error this afternoon and then I will come back to you :)

Cheers
Quentin

**Comment by quentin-cy** [comment]
Hello again @KHRAMS ,

There is indeed an issue with the command `startAirO2ContinuousMeasurement()`, the function does not update the parameters of the conversion function. We will fix it soon and create a new release.

As a temporary workaround, you can start another measurement type beforehand as follows:
```
err = sensor.startO2ContinuousMeasurement();
sensor.stopContinuousMeasurement();
delay(100);
err = sensor.startAirO2ContinuousMeasurement(21);
```
This way the parameters are set by the first command and you will then receive proper readings.

**Comment by KHRAMS** [comment]
Hi @qfisch,

Thank you for dealing with my concerns so swiftly! I'll implement this and get back to you. Also, I may be confused about how these sensors fundamentally work - is it that if I know my gas is a mixture of Air and O2 that I should use this command to get accurate flow readings, or should I use this command to get an understanding of how my oxygen concentration is changing over my mixture flow? Or is it both? Sorry in advance if this is out of scope for this GitHub issue, I am unfortunately not too acquainted with this domain. 

Thank you again!

**Comment by quentin-cy** [comment]
Hello again @KHRAMS ,

Those sensors are calibrated against O2, therefore to get accurate flow readings you need to specify the ratio of O2 in your gas.
Either your have pure O2, Air, or some Air enriched with O2.
The 3 commands tell the sensor in which use-case you are so that it can return accurate reading.
Hope it is clearer now.

Cheers
Q

**Comment by KHRAMS** [comment]
Hey! Thanks for getting back to me again - it is much clearer now. I've actually gone ahead to make a pull-request which solves this original issue. It's quite an easy fix, and doesn't break on my devices.

**Comment by quentin-cy** [comment]
Issue is solved in version [3.0.0](https://github.com/Sensirion/arduino-i2c-sfm-sf06/releases/tag/3.0.0)
Thank you @KHRAMS for your contribution !

Q
